src/base_alloc/base_alloc_linear.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+#include <stdint.h>
+
+#include "base_alloc_internal.h"
+#include "base_alloc_linear.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+
+// minimum size of a single pool of the linear base allocator
+#define MINIMUM_LINEAR_POOL_SIZE (ba_os_get_page_size())
+
+// alignment of the linear base allocator
+#define MEMORY_ALIGNMENT (8)
+
+typedef struct umf_ba_next_linear_pool_t umf_ba_next_linear_pool_t;
+
+// metadata is set and used only in the main (the first) pool
+struct umf_ba_main_linear_pool_meta_t {
+    size_t pool_size; // size of each pool (argument of each ba_os_alloc() call)
+    os_mutex_t *lock;
+    char *data_ptr;
+    size_t size_left;
+#ifndef NDEBUG
+    size_t n_pools;
+#endif /* NDEBUG */
+};
+
+// the main pool of the linear base allocator (there is only one such pool)
+struct umf_ba_linear_pool_t {
+    // address of the beginning of the next pool (a list of allocated pools to be freed in umf_ba_linear_destroy())
+    umf_ba_next_linear_pool_t *next_pool;
+
+    // metadata is set and used only in the main (the first) pool
+    struct umf_ba_main_linear_pool_meta_t metadata;
```

Comment:
After this approach:
`typedef struct {} umf_ba_next_linear_pool_t;`
here is enough:
`umf_ba_main_linear_pool_meta_t metadata;`

Reply:
Done